File: be/KLTN_LibraSys/src/main/java/com/spkt/libraSys/service/dashboard/DashboardService.java
package com.spkt.libraSys.service.dashboard;

import java.util.Map;
import java.time.LocalDate;

/**
 * Service interface for dashboard-related operations
 */
public interface DashboardService {
    
    /**
     * Get document statistics for dashboard
     * @return Map containing document statistics including:
     * - totalDocuments: Total number of documents
     * - documentsByStatus: Number of documents by each status
     * - documentsByType: Number of documents by each type
     * - documentsByCategory: Number of documents by each category
     */
    Map<String, Object> getDocumentStatistics();
    
    /**
     * Get loan statistics for dashboard
     * @return Map containing loan statistics including:
     * - totalLoans: Total number of loans
     * - activeLoans: Number of active loans
     * - overdueLoans: Number of overdue loans
     * - recentLoans: Number of loans in last 30 days
     * - loansByStatus: Number of loans by each status
     */
    Map<String, Object> getLoanStatistics();
    
    /**
     * Get user statistics for dashboard
     * @return Map containing user statistics including:
     * - totalUsers: Total number of users
     * - activeUsers: Number of active users
     * - newUsers: Number of new users in last 30 days
     * - usersByRole: Number of users by each role
     * - usersByStatus: Number of users by each status
     */
    Map<String, Object> getUserStatistics();
    
    /**
     * Get DRM license statistics for dashboard
     * @return Map containing DRM license statistics including:
     * - totalLicenses: Total number of licenses
     * - activeLicenses: Number of active licenses
     * - revokedLicenses: Number of revoked licenses
     * - recentLicenses: Number of licenses in last 30 days
     * - licensesByStatus: Number of licenses by each status
     */
    Map<String, Object> getDrmStatistics();

    /**
     * Get fine statistics for dashboard
     * @return Map containing fine statistics including:
     * - totalFines: Total amount of fines
     * - paidFines: Amount of paid fines
     * - pendingFines: Amount of pending fines
     * - totalFineTransactions: Total number of fine transactions
     * - paidTransactions: Number of paid transactions
     * - pendingTransactions: Number of pending transactions
     * - finesByMonth: Fine amounts by month
     */
    Map<String, Object> getFineStatistics();

    /**
     * Get payment statistics for dashboard
     * @return Map containing payment statistics including:
     * - totalPayments: Total number of payments
     * - cashPayments: Number of cash payments
     * - vnpayPayments: Number of VNPAY payments
     * - totalAmount: Total payment amount
     * - cashAmount: Cash payment amount
     * - vnpayAmount: VNPAY payment amount
     * - paymentsByMonth: Payment amounts by month
     * - paymentMethodDistribution: Distribution of payment methods
     */
    Map<String, Object> getPaymentStatistics();

    /**
     * Get daily statistics for dashboard
     * @return Map containing daily statistics including:
     * - newLoans: Number of new loans today
     * - returns: Number of returns today
     * - newFines: Number of new fines today
     * - newFineAmount: Amount of new fines today
     * - payments: Number of payments today
     * - paymentAmount: Amount of payments today
     * - activeUsers: Number of active users today
     */
    Map<String, Object> getDailyStatistics();

    /**
     * Get monthly statistics for dashboard
     * @return Map containing monthly statistics including:
     * - loansByMonth: Number of loans by month
     * - returnsByMonth: Number of returns by month
     * - finesByMonth: Fine amounts by month
     * - paymentsByMonth: Payment amounts by month
     * - usersByMonth: Number of new users by month
     */
    Map<String, Object> getMonthlyStatistics();

    /**
     * Get yearly statistics for dashboard
     * @return Map containing yearly statistics including:
     * - loansByYear: Number of loans by year
     * - returnsByYear: Number of returns by year
     * - finesByYear: Fine amounts by year
     * - paymentsByYear: Payment amounts by year
     * - usersByYear: Number of new users by year
     */
    Map<String, Object> getYearlyStatistics();
    
    /**
     * Get all dashboard statistics in a single call
     * @return Map containing all dashboard statistics including:
     * - documents: Document statistics
     * - loans: Loan statistics
     * - users: User statistics
     * - drm: DRM statistics
     * - fines: Fine statistics
     * - payments: Payment statistics
     * - daily: Daily statistics
     * - monthly: Monthly statistics
     * - yearly: Yearly statistics
     */
    Map<String, Object> getAllStatistics();

    /**
     * Get document statistics by type
     * @return Map containing document statistics by type including:
     * - totalByType: Total number of documents by each type
     * - typeDistribution: Distribution of documents by type
     */
    Map<String, Object> getDocumentTypeStatistics();

    /**
     * Get loan statistics for a specific date range
     * @param startDate Start date of the range
     * @param endDate End date of the range
     * @return Map containing loan statistics including:
     * - totalLoans: Total number of loans in the period
     * - totalReturns: Total number of returns in the period
     * - activeLoans: Number of active loans
     * - overdueLoans: Number of overdue loans
     * - loansByStatus: Number of loans by each status
     * - dailyLoans: Number of loans per day
     * - dailyReturns: Number of returns per day
     */
    Map<String, Object> getLoanStatisticsByDateRange(LocalDate startDate, LocalDate endDate);
}

File: fe/library_manager/app/services/dashboardService.ts
import apiService from '../untils/api';

interface ApiResponse<T> {
  code: number;
  success: boolean;
  message: string;
  data: T;
}

export interface DashboardStatistics {
  documents: {
    documentsByENABLED: number;
    totalDocuments: number;
    documentsByDISABLED: number;
    typeDistribution?: Record<string, number>;
    totalByType?: Record<string, number>;
  };
  loans: {
    totalLoans: number;
    recentLoans: number;
    activeLoans: number;
    overdueLoans: number;
  };
  fines: {
    pendingTransactions: number;
    pendingFines: number;
    totalFineTransactions: number;
    totalFines: number;
    paidFines: number;
    paidTransactions: number;
  };
  payments: {
    totalAmount: number;
    vnpayPayments: number;
    vnpayAmount: number;
    cashAmount: number;
    cashPayments: number;
    totalPayments: number;
  };
  daily: {
    payments: number;
    newLoans: number;
    returns: number;
    newFines: number;
    newFineAmount: number;
    paymentAmount: number;
  };
  users: {
    totalUsers: number;
    activeUsers: number;
    newUsers: number;
    roleDistribution: {
      role: string;
      count: number;
    }[];
  };
  drm: {
    recentLicenses: number;
    totalLicenses: number;
    revokedLicenses: number;
    activeLicenses: number;
  };
}

const dashboardService = {
  getAllStatistics: async (): Promise<DashboardStatistics> => {
    const response = await apiService.get<ApiResponse<DashboardStatistics>>('/api/dashboard');
    return response.data.data;
  },

  getDocumentStatistics: async (): Promise<DashboardStatistics['documents']> => {
    const response = await apiService.get<any>('/api/dashboard/documents/by-type');
    if (Array.isArray(response.data.data)) {
      return response.data.data[0];
    }
    return response.data.data;
  },

  getLoanStatistics: async (): Promise<DashboardStatistics['loans']> => {
    const response = await apiService.get<ApiResponse<DashboardStatistics['loans']>>('/api/dashboard/loans');
    return response.data.data;
  },

  async getUserStatistics(): Promise<DashboardStatistics['users']> {
    try {
      const response = await apiService.get<{ data: { content: any[] } }>('/api/v1/users');
      const users = response.data.data.content;
      
      // Tính toán thống kê
      const totalUsers = users.length;
      const activeUsers = users.filter(user => user.isActive === 'ACTIVE').length;
      const newUsers = users.filter(user => {
        const createdDate = new Date(user.createdAt);
        const now = new Date();
        const diffDays = Math.floor((now.getTime() - createdDate.getTime()) / (1000 * 60 * 60 * 24));
        return diffDays <= 7; // Người dùng mới trong 7 ngày
      }).length;

      // Tính toán phân bố roles
      const roleCounts = users.reduce((acc: { [key: string]: number }, user) => {
        user.roles.forEach((role: string) => {
          acc[role] = (acc[role] || 0) + 1;
        });
        return acc;
      }, {});

      const roleDistribution = Object.entries(roleCounts).map(([role, count]) => ({
        role,
        count
      }));

      return {
        totalUsers,
        activeUsers,
        newUsers,
        roleDistribution
      };
    } catch (error) {
      console.error('Error fetching user statistics:', error);
      throw error;
    }
  },

  getDrmStatistics: async (): Promise<DashboardStatistics['drm']> => {
    const response = await apiService.get<ApiResponse<DashboardStatistics['drm']>>('/api/dashboard/drm');
    return response.data.data;
  },

  getFineStatistics: async (): Promise<DashboardStatistics['fines']> => {
    const response = await apiService.get<ApiResponse<DashboardStatistics['fines']>>('/api/dashboard/fines');
    return response.data.data;
  },

  getPaymentStatistics: async (): Promise<DashboardStatistics['payments']> => {
    const response = await apiService.get<ApiResponse<DashboardStatistics['payments']>>('/api/dashboard/payments');
    return response.data.data;
  },

  getDailyStatistics: async (): Promise<DashboardStatistics['daily']> => {
    const response = await apiService.get<ApiResponse<DashboardStatistics['daily']>>('/api/dashboard/daily');
    return response.data.data;
  },

  getDocumentStatusStatistics: async (): Promise<DashboardStatistics['documents']> => {
    const response = await apiService.get<ApiResponse<DashboardStatistics['documents']>>('/api/dashboard/documents');
    return response.data.data;
  },

  getDocumentTypeStatistics: async (): Promise<{ typeDistribution?: Record<string, number>; totalByType?: Record<string, number>; }> => {
    const response = await apiService.get<any>('/api/dashboard/documents/by-type');
    if (Array.isArray(response.data.data)) {
      return response.data.data[0];
    }
    return response.data.data;
  },

  getLoanStatisticsByDateRange: async (startDate: string, endDate: string) => {
    const response = await apiService.get<any>(`/api/dashboard/loans/statistics/date-range?startDate=${startDate}&endDate=${endDate}`);
    return response.data.data;
  },

//   getMonthlyStatistics: async () => {
//     const response = await apiService.get('/api/dashboard/monthly');
//     return response.data.data;
//   },

//   getYearlyStatistics: async () => {
//     const response = await apiService.get('/api/dashboard/yearly');
//     return response.data.data;
//   },
};

export default dashboardService; 